src/cs/Ssh/SecureStream.cs

```diff
@@ -128,50 +129,60 @@ public async Task ConnectAsync(CancellationToken cancellation = default)
 				"An Authenticating event handler must be registered before connecting.");
 		}
 
-		if (this.serverCredentials != null)
+		try
 		{
-			var serverSession = (SshServerSession)this.session;
-			serverSession.Credentials = this.serverCredentials;
-		}
-
-		await this.session.ConnectAsync(this.transportStream, cancellation).ConfigureAwait(false);
-
-		SshChannel? channel = null;
-		if (this.clientCredentials != null)
-		{
-			SshConnectionException? ex = null;
-			var clientSession = (SshClientSession)this.session;
-			if (!(await clientSession.AuthenticateServerAsync(cancellation).ConfigureAwait(false)))
+			if (this.serverCredentials != null)
 			{
-				ex = new SshConnectionException(
-					"Server authentication failed.", SshDisconnectReason.HostKeyNotVerifiable);
+				var serverSession = (SshServerSession)this.session;
+				serverSession.Credentials = this.serverCredentials;
 			}
 
-			if (ex == null && !(await clientSession.AuthenticateClientAsync(
-				this.clientCredentials, cancellation).ConfigureAwait(false)))
+			await this.session.ConnectAsync(this.transportStream, cancellation).ConfigureAwait(false);
+
+			SshChannel? channel = null;
+			if (this.clientCredentials != null)
 			{
-				ex = new SshConnectionException(
-					"Client authentication failed.", SshDisconnectReason.NoMoreAuthMethodsAvailable);
+				SshConnectionException? ex = null;
+				var clientSession = (SshClientSession)this.session;
+				if (!(await clientSession.AuthenticateServerAsync(cancellation).ConfigureAwait(false)))
+				{
+					ex = new SshConnectionException(
+						"Server authentication failed.", SshDisconnectReason.HostKeyNotVerifiable);
+				}
+
+				if (ex == null && !(await clientSession.AuthenticateClientAsync(
+					this.clientCredentials, cancellation).ConfigureAwait(false)))
+				{
+					ex = new SshConnectionException(
+						"Client authentication failed.", SshDisconnectReason.NoMoreAuthMethodsAvailable);
+				}
+
+				if (ex != null)
+				{
+					await this.session.CloseAsync(ex.DisconnectReason, ex).ConfigureAwait(false);
+					throw ex;
+				}
+
+				channel = await this.session.OpenChannelAsync(cancellation).ConfigureAwait(false);
 			}
-
-			if (ex != null)
+			else
 			{
-				await this.session.CloseAsync(ex.DisconnectReason, ex).ConfigureAwait(false);
-				throw ex;
+				channel = await this.session.AcceptChannelAsync(cancellation).ConfigureAwait(false);
 			}
 
-			channel = await this.session.OpenChannelAsync(cancellation).ConfigureAwait(false);
+			this.stream = CreateStream(channel);
+			channel.Closed += (_, _) =>
+			{
+				this.Dispose();
+			};
 		}
-		else
+		catch (Exception ex)
 		{
-			channel = await this.session.AcceptChannelAsync(cancellation).ConfigureAwait(false);
+			this.connectCompletion.TrySetException(ex);
```

Comment:
Should we close the session here too? We do close it if authentication fails above.

Reply:
Updated.